### ISSUE TITLE
arbitrary order of inputs/parameters

### ISSUE THREAD

**Issue by jairideout** [qiime2/qiime2#70]
Noticed this while playing around with q2cli today. When calling `--help` on a method or visualizer, the order of inputs and parameters changes across runs, which is going to be confusing to users. This doesn't happen with methods defined in Markdown files because order is preserved in the Markdown file, and `Method.from_markdown` uses `OrderedDict` internally to preserve that order.

We have a couple of options:

1. Display all options in alphabetical order in q2cli and qiime-studio, regardless of underlying function signature. Technically no change is necessary to the framework if we go this route.
2. Display options in the order they are defined in the function signature. Note that this won't always group input artifacts followed by parameters, as plugin developers can define their function signature in whatever order they wish. The plugin developer understands the ordering of parameters best, so perhaps it's best if we respect their defined ordering in (all?) interfaces. If we go this route, we'll need to update the framework to preserve order, and q2cli probably won't need updating because it's just looping over the `signature` (unsure about qiime-studio).

@gregcaporaso @ebolyen @jakereps what do you think? I'm leaning towards option 2.

**Comment by ebolyen** [comment]
I also like option 2, however this would mean `Signature` needs something (like `__iter__`!) to describe how artifacts/parameters are intermingled as their separated forms won't work. 

That being said I would rather see all Artifact inputs in the front and all parameters in the back. This "front loads" you with the important bits first and the little details later (especially if optional Artifacts aren't allowed). Mixing the two just sounds like a confusing way to structure your API, it feels the same as intermingling positional and optional parameters. But that is just my opinion (for the moment).

**Comment by jakereps** [comment]
Oh, to clarify, is this about cross-category ordering, where say a parameter could come between two inputs, or within each of the three separate categories? I'd vote option 2 would be best, as some things out of the functional/defined order could seem strange, and either way it would definitely need some sort of order map accessible to the studio/JS side, as once it leaves qiime, even if in an OrderedDict when sent through JSON, there's no way of assuring it stays how it was on the React side.

Right now the studio sorts plugins and methods alphabetically, as previously at every launch, and even during runs, they would shuffle around (during runs if something was updated (required input artifact list or something) and they were redrawn), so if it ends up just being an alphabetical sort within each category, like option 1, it would require no "real" changes, but given some sort of order definition it shouldn't be too much to get option 2 working for the studio as well.

**Comment by jairideout** [comment]
How about going with option 2 (preserving order of parameters, as the plugin dev knows best) and at registration time we disallow function signatures that intermingle artifacts and parameters. IOW, we require a function to have input artifacts first followed by parameters. @ebolyen does that solve your concern? Then `Signature` won't need updating, we just update `from_function` to preserve order when it constructs the `Signature`.

@jakereps for the studio we should be able to preserve order even without `OrderedDict` because a list of tuples (or list of lists) can be shipped across the wire. Something like:

```python
[('artifact1', (SemanticType, view_type)), ('param1', (PrimitiveType, view_type)), ...]
```

This structure is actually one of the ways to construct an `OrderedDict`. Does that work for the studio?

**Comment by jakereps** [comment]
👍 That should work great, @jairideout.

**Comment by ebolyen** [comment]
:+1: That should work great, both explicit and well structured.

**Comment by jairideout** [comment]
Temporary fix is in for q2cli (https://github.com/qiime2/q2cli/pull/33) which sorts options alphabetically within input, parameter, and output groups. Revert q2cli commit 9802193c7a8fa6b7537bc9565712d95e3959e9cd when this issue is fixed to retain function signature order.

**Comment by jairideout** [comment]
Fixed in #170.
